### PR TITLE
Upgrade to go 1.18

### PR DESCRIPTION
**Pull request by izaaklauer**
This PR updates us to go 1.18. 

Closes https://github.com/hashicorp/waypoint/pull/3284

### CODE REVIEW

File: nix/overlay.nix [hunk 1]
```diff
@@ -5,7 +5,7 @@ final: prev: {
   devShell = final.callPackage ./waypoint.nix { };
 
   # Need to manually do this since 1.16 is still the default
-  go = final.go_1_17;
+  go = final.go_1_18;
```
Comment (izaaklauer) [inline nix/overlay.nix:8]: @mitchellh Let me know if this isn't correct

Comment (mitchellh) [reply nix/overlay.nix:8]: Probably not, because we pin the index so 1.18 is prob not available. Just leave it out and I'll upgrade myself later. :)

**Comment by izaaklauer** [comment]
Lets hold off on this for now - it looks like golang-ci doesn't fully support go 1.18 yet: https://github.com/golangci/golangci-lint/issues/2649

**Comment by izaaklauer** [comment]
Superseded by the upgrade to go 1.19.

**Pull request closed without merging**